File: src/test/java/org/trp/shincolle/init/EntityLocalizationRegressionTest.java
package org.trp.shincolle.init;

import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertTrue;

class EntityLocalizationRegressionTest {
    private static final Path MOD_ENTITIES_SOURCE =
            Path.of("src/main/java/org/trp/shincolle/init/ModEntities.java");
    private static final List<Path> LANGUAGE_SOURCES = List.of(
            Path.of("src/main/resources/assets/shincolle/lang/en_us.json"),
            Path.of("src/main/resources/assets/shincolle/lang/ja_jp.json"),
            Path.of("src/main/resources/assets/shincolle/lang/zh_cn.json"),
            Path.of("src/main/resources/assets/shincolle/lang/zh_tw.json")
    );
    private static final List<String> ENTITY_KEYS = List.of(
            "entity.shincolle.abyss_missile",
            "entity.shincolle.carrier_w_demon",
            "entity.shincolle.mount_is_h",
            "entity.shincolle.mount_mi_h",
            "entity.shincolle.mount_su_h",
            "entity.shincolle.projectile_beam",
            "entity.shincolle.ship_fishing_hook",
            "entity.shincolle.ship_grudge",
            "entity.shincolle.takoyaki"
    );

    @Test
    void registeredEntityNamesShouldRemainLocalizedInMaintainedLanguages() throws IOException {
        String modEntities = Files.readString(MOD_ENTITIES_SOURCE);

        for (String key : ENTITY_KEYS) {
            assertRegisteredEntitySourceStillUses(key, modEntities);
            assertLocalizedInMaintainedLanguages(key);
        }
    }

    private static void assertRegisteredEntitySourceStillUses(String key, String modEntities) {
        String entityId = key.substring("entity.shincolle.".length());
        assertTrue(modEntities.contains("register(\"" + entityId + "\""),
                () -> "Expected ModEntities to keep registering entity " + entityId);
    }

    private static void assertLocalizedInMaintainedLanguages(String key) throws IOException {
        for (Path languageSource : LANGUAGE_SOURCES) {
            String source = Files.readString(languageSource);
            assertTrue(source.contains("\"" + key + "\""),
                    () -> "Expected maintained languages to define " + key);
        }
    }
}

File: src/test/kotlin/org/trp/shincolle/init/EntityLocalizationRegressionTest.kt
package org.trp.shincolle.init

import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.nio.file.Files
import java.nio.file.Path
import java.util.regex.Pattern

class EntityLocalizationRegressionTest {
    @Test
    fun registeredEntityNamesShouldRemainLocalizedInMaintainedLanguages() {
        val modEntities = Files.readString(MOD_ENTITIES_SOURCE)
        val entityKeys = readRegisteredEntityIds(modEntities).map { entityId -> "entity.shincolle.$entityId" }.toSet()

        assertTrue(entityKeys.isNotEmpty())

        entityKeys.forEach { key ->
            assertLocalizedInMaintainedLanguages(key)
        }
    }

    private fun readRegisteredEntityIds(modEntities: String): List<String> {
        val entityIds = mutableListOf<String>()
        val matcher = ENTITY_REGISTRATION_PATTERN.matcher(modEntities)
        while (matcher.find()) {
            entityIds += matcher.group(1)
        }
        return entityIds
    }

    private fun assertLocalizedInMaintainedLanguages(key: String) {
        LANGUAGE_SOURCES.forEach { languageSource ->
            val languageMap = readLanguageMap(languageSource)
            assertTrue(languageMap.containsKey(key)) {
                "Expected ${languageSource.fileName} to define $key"
            }
        }
    }

    private fun readLanguageMap(path: Path): Map<String, String> {
        val content = Files.readString(path)
        val map = mutableMapOf<String, String>()
        val matcher = LANGUAGE_ENTRY_PATTERN.matcher(content)
        while (matcher.find()) {
            map[matcher.group(1)] = matcher.group(2)
        }
        return map
    }

    companion object {
        private val MOD_ENTITIES_SOURCE: Path = Path.of("src/main/java/org/trp/shincolle/init/ModEntities.kt")
        private val LANGUAGE_SOURCES = listOf(
            Path.of("src/main/resources/assets/shincolle/lang/en_us.json"),
            Path.of("src/main/resources/assets/shincolle/lang/ja_jp.json"),
            Path.of("src/main/resources/assets/shincolle/lang/zh_cn.json"),
            Path.of("src/main/resources/assets/shincolle/lang/zh_tw.json")
        )
        private val ENTITY_REGISTRATION_PATTERN: Pattern = Pattern.compile(
            "ENTITY_TYPES\\.register<.*?>\\(\\s*\"([a-z0-9_]+)\"",
            Pattern.DOTALL
        )
        private val LANGUAGE_ENTRY_PATTERN: Pattern = Pattern.compile("\"([^\"]+)\"\\s*:\\s*\"((?:[^\"\\\\]|\\\\.)*)\"")
    }
}
